control-plane/addMissingIds.cpp

```diff
@@ -0,0 +1,214 @@
+#include "addMissingIds.h"
+
+#include "frontends/p4/evaluator/evaluator.h"
+#include "p4RuntimeAnnotations.h"
+#include "p4RuntimeArchStandard.h"
+
+namespace P4 {
+
+const IR::P4Program* MissingIdAssigner::preorder(IR::P4Program* program) {
+    auto evaluator =  P4::EvaluatorPass(refMap, typeMap);
+    program->apply(evaluator);
+    auto* toplevel = evaluator.getToplevelBlock();
+    CHECK_NULL(toplevel);
+    symbols = ControlPlaneAPI::P4RuntimeSymbolTable::generateSymbols(
+        toplevel->getProgram(), toplevel, refMap, typeMap,
+        archBuilder(refMap, typeMap, toplevel));
+    auto objectVector = IR::Vector<IR::Node>();
+    for (const auto* obj : program->objects) {
+        objectVector.push_back(obj->apply(
+            MissingIdAssigner(refMap, typeMap, symbols, archBuilder)));
+    }
+    program->objects = objectVector;
+    return program;
+}
+
+const IR::P4Table* MissingIdAssigner::preorder(IR::P4Table* table) {
+    if (ControlPlaneAPI::isHidden(table)) {
+        return table;
+    }
+    CHECK_NULL(symbols);
+    auto anno = ControlPlaneAPI::getIdAnnotation(table);
+    if (!anno) {
+        const auto* annos = table->getAnnotations();
+        auto* newAnnos = annos->clone();
+        IR::Vector<IR::Expression> annoExprs;
+        auto symbolId = symbols->getId(
+            ControlPlaneAPI::P4RuntimeSymbolType::TABLE(), table);
+        const auto* idConst =
+            new IR::Constant(new IR::Type_Bits(ID_BIT_WIDTH, false), symbolId);
+        annoExprs.push_back(idConst);
+        newAnnos->add(new IR::Annotation("id", annoExprs));
+        table->annotations = newAnnos;
+    }
+    size_t propertyIdx = 0;
+
+    const IR::Key* key = nullptr;
```

Comment:
you are reconstructing here the logic of the visitor, why not just make a postorder function for a table property?